# Preference of Online Music Platform Customers

# Contents <a id='back'></a>

* [Introduction](#intro)
* [1. Data Review](#data_review)
    * [Data Review Conclusions](#data_review_conclusions)
* [2. Data Preprocessing](#data_preprocessing)
    * [2.1 Header Style](#header_style)
    * [2.2 Missing Values](#missing_values)
    * [2.3 Duplicates](#duplicates)
    * [2.4 Data Preprocessing Conclusions](#data_preprocessing_conclusions)
* [3. Hypotheses](#hypotheses)
    * [3.1 Hypothesis 1: Activity](#activity)
    * [3.2 Hypotheses 2: Week](#week)
    * [3.3 Hypotheses 3: Genre](#genre)
* [Findings](#end)

# Introduction <a id='intro'></a>

This report provides a comparison overview of music preference across listeners in Springfield and Shelbyville, involving data owned by Y.Music regarding customer behavior. 

Objectives: 
To test hypotheses as follow:
1. Customer activities differ based on day and city. 
2. There is a difference in music genre that customers listen to on Monday AM across the two cities. Same applies to Friday PM. 
3. Springfield listeners prefer pop, while Shelbyville listeners prefer rap. 

Data Analysis
The project analysis involves:
 1. Data review
 2. Data Preprocessing
 3. Hypothesis test
 


## Data Review <a id='data_review'></a>

In [1]:
# import library
import pandas as pd

In [2]:
# save files in dataframe 'df'
df = pd.read_csv('/datasets/music_project_en.csv')

# display dataset
df.describe()

userID  Track     artist  genre       City        time     Day
count      65079  63736      57512  63881        65079     65079   65079
unique     41748  39666      37806    268            2     20392       3
top     A8AE9169  Brand  Kartvelli    pop  Springfield  08:14:07  Friday
freq          76    136        136   8850        45360        14   23149

In [3]:
# first 10 lines
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

In [4]:
df.info()# general information of df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Table contains 7 (seven) columns, all in the same data type ('object'). 

According to documentation:
- `'userID'` — User ID number
- `'Track'` — title of the track
- `'artist'` — artist's name
- `'genre'` - genre
- `'City'` — city where listener is located
- `'time'` — the time track was being played
- `'Day'` — name of days in a week

Problems in data that needs fixing:
1. Inconsistency of upper and lower case.
2. Inconsistency of usage of spacing between words
3. Column name merging two words, suppossedly separated by an underscore '_'.

Differences of non-null values across columns indicates some missing values. 


### Data Review Conclusions  <a id='data_review_conclusions'></a> 

Each row in the table contains data regarding played tracks. Some columns contain data that describes the track: title, artist, and genre. The others contain info regarding user: city, and time where track was being played. 

There are also missing values in the dataset, which will be addressed in the following section (Data Preprocessing)

[Back to Contents](#back)

## Data Preprocessing <a id='data_preprocessing'></a>

### Header Style <a id='header_style'></a>

In [5]:
print(df.columns)# display list of column names

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


In [6]:
df = df.rename(columns={
    '  userID' : 'user_id',
    'Track' : 'track',
    '  City  ' : 'city',
    'Day' : 'day'})# renaming column names

Column names after revision:

In [7]:
df.head(10)

user_id                        track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

          city      time        day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

[Back to Contents](#back)

### Missing Values <a id='missing_values'></a>

In [8]:
df.isna().sum()# show the count of missing values
df.head()

user_id                        track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

          city      time        day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday

Only missing values in column 'genre' will affect the data analysis. 
Missing values in column 'track' and 'artist' can be replaced by string 'unknown'. 

Missing values in column 'track', 'artist' and 'genre' will be replaced by a string named 'unknown'. Only values in column 'genre' will affect the result of this analysis, hence further evaluation will be carried to learn the proportion of 'unknown' rows in column genre. 

In [9]:
columns_to_replace = ['track', 'artist', 'genre']
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')

df.head()
                      
    # apply loop 'for' to column with missing values, and replace with 'unknown'

user_id                        track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

          city      time        day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday

In [10]:
df.isna().sum()# display count of missing values after loop 

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

There is no missing value accross column.

[Back to Table of contents](#back)

### Duplicates <a id='duplicates'></a>

In [11]:
df.duplicated().sum()# display count of explicit duplicates

3826

There are 3826 duplicates in the datasets. Proceed to drop duplicates:

In [12]:
df = df.drop_duplicates().reset_index(drop=True)# drop explicit duplicates

In [13]:
df.duplicated().sum()# checking duplicates

0

The following is to check implicit duplicate in column 'genre', e.g. the same genre but written in different combination of upper and lowercase. 

In [14]:
sorted_by_genre = df.sort_values(by='genre')
sorted_by_genre

user_id                       track  \
31916  C40BC5DF            Shot in the Dark   
3636   F10919ED          Ready For The Fire   
20264  B02ED3AB  Prelude – Ne Me Qui Te Pas   
37784  77D3A7D9  Prelude – Ne Me Qui Te Pas   
14611  1C7D9E90  Prelude – Ne Me Qui Te Pas   
...         ...                         ...   
6158   12A10DD5                     Ekatana   
17963   653E0EF                     Kalinka   
27939  D7FB50DA             Drumming Circle   
39063  AA1730E8                         Anu   
8448   A439123F                     Flip It   

                                                  artist      genre  \
31916                                         Four80East       acid   
3636                                    Valley Of Wolves   acoustic   
20264                                            unknown   acoustic   
37784                                            unknown   acoustic   
14611                                            unknown   acoustic   
...                                                  ...        ...   
6158                                         Shantavaani      world   
17963  The Red Army Choirs Of Alexandrov (Les Choeurs...      world   
27939                                   Professor Trance  worldbeat   
39063                                  Kailash Kokopelli  worldbeat   
8448                                             unknown        ïîï   

              city      time        day  
31916  Springfield  09:44:12     Friday  
3636   Shelbyville  21:33:18     Monday  
20264  Springfield  13:31:03  Wednesday  
37784  Springfield  14:27:34     Monday  
14611  Shelbyville  21:47:42     Friday  
...            ...       ...        ...  
6158   Springfield  14:30:19     Friday  
17963  Springfield  21:08:36  Wednesday  
27939  Springfield  09:30:47     Monday  
39063  Shelbyville  20:16:34  Wednesday  
8448   Springfield  09:08:51     Friday  

[61253 rows x 7 columns]

In [15]:
sorted_by_genre['genre'].unique() #display unique values in column 'genre' in alphabetical order

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'ambient', 'americana', 'animated', 'anime',
       'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook',
       'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom',
       'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian',
       'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop',
       'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber',
       'children', 'chill', 'chinese', 'choral', 'christian', 'christmas',
       'classical', 'classicmetal', 'club', 'colombian', 'comedy',
       'conjazz', 'contemporary', 'country', 'cuban', 'dance',
       'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock',
       'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat',
       'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy',
       'electronic', 'electropop', 'emo', 'entehno', 'epicmetal',
       'estrada', 'ethnic', 'eurofo

The list shows some implicit duplicate for 'hiphop' genre, where in some rows it was input as 'hiphop', 'hip-hop', 'hip' and 'hop'. 

Implicit duplicates will be addressed with function 'replace_wrong_genres' using two parameters:
* `wrong_genres=` — list containing duplicates that need to be changed
* `correct_genre=` — string of correct value


In [16]:
def replace_wrong_genres(wrong_genres, correct_genres):
    sorted_by_genre['genre'] = sorted_by_genre['genre'].replace(wrong_genres, correct_genres)
wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genres = 'hiphop'
# function to correct implicit duplicates

In [17]:
replace_wrong_genres(wrong_genres, correct_genres)
sorted_by_genre['genre']# call function 

31916         acid
3636      acoustic
20264     acoustic
37784     acoustic
14611     acoustic
           ...    
6158         world
17963        world
27939    worldbeat
39063    worldbeat
8448           ïîï
Name: genre, Length: 61253, dtype: object

Checking if all implicit duplicates have been replaced:

In [18]:
# display unique values of 'genre'
sorted_by_genre['genre'].unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'ambient', 'americana', 'animated', 'anime',
       'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook',
       'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom',
       'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian',
       'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop',
       'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber',
       'children', 'chill', 'chinese', 'choral', 'christian', 'christmas',
       'classical', 'classicmetal', 'club', 'colombian', 'comedy',
       'conjazz', 'contemporary', 'country', 'cuban', 'dance',
       'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock',
       'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat',
       'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy',
       'electronic', 'electropop', 'emo', 'entehno', 'epicmetal',
       'estrada', 'ethnic', 'eurofo

[Back to Table of Contents](#back)

### Data Preprocessing Conclusins <a id='data_preprocessing_conclusions'></a>

There has been three problems identified in the dataset:

- Inconsistent column naming
- Missing values
- Explicit and implicit duplicates

Column names have been uniformed to all use lowercase and underscore to separate multiple words. 
All missing values have been replaced with string 'unknown'. 
Explicit duplicates in column 'track' and 'artist' have been replaced by string 'unknown'. And implicit duplicates in column 'genre' have been fixed to uniformed word. 

[Back to Table of Contents](#back)

## Hypotheses <a id='hypotheses'></a>

### Hypothesis 1: Activity <a id='activity'></a>

The first hypothesis stated that listeners of the Springfield and Shelbyville differ in terms of music listening behavior. The hypothesis was based on data of activities taken from three days in a week: Monday, Wednesday, and Friday. 

Hypothesis will be tested with the following steps:
1. Divide users in groups based on cities (group Springfield and group Shelbyville)
2. Compare the count of tracks being played by each group on Monday, Wednesday, and Friday.

In [19]:
df.groupby('city')['track'].count()# Count played tracks from each city

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

Given the larger number of residents in Springfield, the number of played tracks in general is also larger compared to Shelbyville. 

The following is count of played tracks grouped by day (Monday, Wednesday, and Friday)

In [20]:
df.groupby('day')['track'].count() #count of played tracks by day

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

In general, Wednesday is the day with least count of played tracks. 

The following is a function to learn number of tracks played by day and city. 

The function is called `number_tracks()` aiming to calculate the count of tracks played for certain day AND city. The function will involve two parameters as follow: 
* name of day
* came of city

The function will use following variables to contain rows from original dataframe, where
  * Values from column `'day'` is parameter `day`\n",
  * Values from column `'city'` is parameter `city`

In [21]:
# <declare function number_tracks()>
tracks = df.loc[(df['city'] == 'Shelbyville') & (df['day'] == 'Wednesday')]

track_list_count = tracks['user_id'].count()
print(track_list_count)

def number_tracks(city, day):
    tracks = df.loc[(df['city']==city) & (df['day']==day)]
    track_list_count = tracks['user_id'].count()
    return track_list_count

7003


The following are display of number of tracks played in each city according to days, using function `number_tracks()`

In [22]:
# count of played tracks in Springfield on Monday
number_tracks(city='Springfield', day='Monday')

15740

In [23]:
# count of played tracks in Shelbyville on Monday
number_tracks(city='Shelbyville', day='Monday')

5614

In [24]:
# count of played tracks in Springfield on Wednesday
number_tracks(city='Springfield', day='Wednesday')

11056

In [25]:
# count of played tracks in Shelbyville on Wednesday
number_tracks(city='Shelbyville', day='Wednesday')

7003

In [26]:
# count of played tracks in Springfield on Friday
number_tracks(city='Springfield', day='Friday')

15945

In [27]:
# count of played tracks in Shelbyville on Friday
number_tracks(city='Shelbyville', day='Friday')

5895

Compiling the data generated by function into a dataframe called "track_list"

In [28]:
# tabel to store results
table_content = [  
        ['Springfield', number_tracks(city='Springfield', day='Monday'), number_tracks(city='Springfield', day='Wednesday'), number_tracks(city='Springfield', day='Friday') ],  
        ['Shelbyville', number_tracks(city='Shelbyville', day='Monday'), number_tracks(city='Shelbyville', day='Wednesday'), number_tracks(city='Shelbyville', day='Friday')]  
          
]
column_name = ['city', 'monday', 'wednesday', 'friday']

track_list = pd.DataFrame(data=table_content , columns=column_name)
track_list             

city  monday  wednesday  friday
0  Springfield   15740      11056   15945
1  Shelbyville    5614       7003    5895

**Conclusion**

The data indicates differences in customer activities as follow:

- Number of played tracks in Springfield peak on Monday and Friday, and decline on Wednesday
- Number of played tracks in Shelbyville peak on Wednesday, and decline on Monday and Friday. 

Findings from this section fails to reject the first hypothesis. 

[Back to Contents](#back)

### Hypothesis 2: Weekend vs Weekdays <a id='week'></a>

The second hypothesis stated that on Monday AM and Friday PM, listeners from Springfield listened to different music genre from listener from Shelbyville. 

Testing the anaylisis would first require to separate the table according to cities: 
* For Springfield — `spr_general`
* For Shelbyville — `shel_general`

In [29]:
# generate table spr_general from rows in df,
# which has value 'Springfield' in column 'city'
spr_general = df[df['city'] == 'Springfield']
spr_general

user_id                          track                   artist  \
1      55204538    Delayed Because of Accident         Andreas Rönnberg   
4      E2DC1FAE                    Soul People               Space Echo   
6      4CB90AA5                           True             Roman Messer   
7      F03E1C1F               Feeling This Way          Polina Griffith   
8      8FA1D3BE                       L’estate              Julia Dalia   
...         ...                            ...                      ...   
61247  83A474E7  I Worship Only What You Bleed  The Black Dahlia Murder   
61248  729CBB09                        My Name                   McLean   
61250  C5E3A0D5                      Jalopiina                  unknown   
61251  321D0506                  Freight Train            Chas McDevitt   
61252  3A64EF84      Tell Me Sweet Little Lies             Monica Lopez   

              genre         city      time        day  
1              rock  Springfield  14:07:09     Friday  
4             dance  Springfield  08:34:34     Monday  
6             dance  Springfield  13:00:07  Wednesday  
7             dance  Springfield  20:47:49  Wednesday  
8            ruspop  Springfield  09:17:40     Friday  
...             ...          ...       ...        ...  
61247  extrememetal  Springfield  21:07:12     Monday  
61248           rnb  Springfield  13:32:28  Wednesday  
61250    industrial  Springfield  20:09:26     Friday  
61251          rock  Springfield  21:43:59     Friday  
61252       country  Springfield  21:59:46     Friday  

[42741 rows x 7 columns]

In [30]:
# generate table spr_general from rows in df,
# which has value 'Shelbyville' in column 'city'
shel_general = df[df['city'] == 'Shelbyville']
shel_general

user_id                              track              artist  \
0      FFB692EC                  Kamigata To Boots    The Mass Missile   
2        20EC38                  Funiculì funiculà         Mario Lanza   
3      A3DD03C9              Dragons in the Sunset          Fire + Ice   
5      842029A1                             Chains            Obladaet   
9      E772D5C0                          Pessimist             unknown   
...         ...                                ...                 ...   
61239  D94F810B        Theme from the Walking Dead  Proyecto Halloween   
61240  BC8EC5CF       Red Lips: Gta (Rover Rework)               Rover   
61241  29E04611                       Bre Petrunko       Perunika Trio   
61242  1B91C621             (Hello) Cloud Mountain     sleepmakeswaves   
61249  D08D4A55  Maybe One Day (feat. Black Spade)         Blu & Exile   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
5          rusrap  Shelbyville  13:09:41     Friday  
9           dance  Shelbyville  21:20:49  Wednesday  
...           ...          ...       ...        ...  
61239        film  Shelbyville  21:14:40     Monday  
61240  electronic  Shelbyville  21:06:50     Monday  
61241       world  Shelbyville  13:56:00     Monday  
61242    postrock  Shelbyville  09:22:13     Monday  
61249         hip  Shelbyville  10:00:00     Monday  

[18512 rows x 7 columns]

A function called `genre_weekday()` with four parameter:
* A table for data
* Name of day
* Starting time stamp in 'hh:mm' format
* End time stamp in 'hh:mm' format

This function will show information of 15 most popular genre on certain days within the timestamp period. 

In [31]:
   # genre_df will only show rows in df which 'day' column contains the same value with 'day'
genre_df = df.loc[df['day']=='Monday']

    # genre_df will only show rows in df which time stamp is lower than 'time2'
genre_df = genre_df.loc[genre_df['time'] < '12:00:00']

    # genre_df will only show rows in df which time stamp is greater than 'time1'
genre_df = genre_df.loc[genre_df['time'] > '09:00:00']

    # group DataFrame which has been filtered based on column 'genre', and find the number of rows by counting number of users 
genre_df_grouped = genre_df.groupby('genre')['user_id'].count()

    # # sort result in descending order to get most popular displayed first
genre_df_sorted = genre_df_grouped.sort_values(ascending=False)
    

    # Limit to show only 15 most popular genres
genre_df_sorted[:15]

genre
pop            494
dance          378
rock           327
electronic     293
hip            175
ruspop         117
alternative    114
rusrap         113
world          107
classical      101
unknown         92
jazz            75
metal           71
folk            64
soundtrack      64
Name: user_id, dtype: int64

Calling the function to compare result on weekday for Springfield and Shelbyville Bandingkan hasil dari fungsi `genre_weekday()` untuk Springfield dan Shelbyville pada hari Senin pagi (dari pukul 07.00 hingga 11.00) dan pada hari Jumat malam (dari pukul 17:00 hingga 23:00):

In [32]:
# memanggil fungsi untuk Senin pagi di Springfield (gunakan spr_general alih-alih tabel df)
def genre_weekday(dataframe, day, time1, time2):
    genre_df = dataframe.loc[df['day']==day]
    genre_df = genre_df.loc[genre_df['time'] < time2]
    genre_df = genre_df.loc[genre_df['time'] > time1]
    genre_df_grouped = genre_df.groupby('genre')['user_id'].count()
    genre_df_sorted = genre_df_grouped.sort_values(ascending=False)
    return genre_df_sorted[:15]

genre_weekday(spr_general, 'Monday', '07:00:00', '11:00:00')

genre
pop            781
dance          549
electronic     480
rock           474
hip            281
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
classical      157
metal          120
jazz           100
folk            97
soundtrack      95
Name: user_id, dtype: int64

In [33]:
# apply function for Monday AM in Shelbyvilled, refering to table 'shel_general'
genre_weekday(shel_general, 'Monday', '07:00:00', '11:00:00')

genre
pop            218
dance          182
rock           162
electronic     147
hip             79
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
world           36
rap             32
soundtrack      31
rnb             27
metal           27
Name: user_id, dtype: int64

In [34]:
# apply function to Friday PM for Springfield refering to table 'spr_general'
genre_weekday(spr_general, 'Friday', '17:00:00', '23:00:00')

genre
pop            713
rock           517
dance          495
electronic     482
hip            267
world          208
ruspop         170
classical      163
alternative    163
rusrap         142
jazz           111
unknown        110
soundtrack     105
rnb             90
metal           88
Name: user_id, dtype: int64

In [35]:
# apply function to Friday PM in Shelbyville
genre_weekday(shel_general, 'Friday', '17:00:00', '23:00:00')

genre
pop            256
electronic     216
rock           216
dance          210
hip             94
alternative     63
jazz            61
classical       60
rusrap          59
world           54
unknown         47
ruspop          47
soundtrack      40
metal           39
rap             36
Name: user_id, dtype: int64

**Conclusion**

Comparison of top 15 genres on Monday AM in both cities has concluded the following: 

1. Listeners in Springfield and Shelbyville share similar genre preferences. The top five list of both cities contain similar genres, and in relatively similar order with an exception of Rock and Electronic.

2. There is a considerably significant proportion of missing values of track genre in the dataset of Springfield listeners, making 'unknown' in the top 10 of the list. Investigation of the missing value is highly suggested to ensure reliability of the findings. 

Similar findings apply for Friday PM. Individual genre varies, however in general both cities share similar preferences regarding top 15 genres.

With regards to hypotheses: 
* "Customers listen to similar music on weekdays and weekend": correct. 
* "Listeners from Springfield listened to different music genre from listener from Shelbyville": incorrect. 

[Back to Contents](#back)

### Hypothesis 3: Genre <a id='genre'></a>

Hypothesis: Listeners from Sheblyville prefers Rap, while listeners from Springfield prefers Pop.

In order to test this hypothesis, the following steps are required:

First to group table `spr_general` by genre, and find the played tracks count for each genre with `count()`. Results then will be stored in `spr_genres` and displayed in descending order, with the top row being the most popular genre in Springfield. 

In [36]:
# grouping table spr_general by column 'genre',
spr_general.groupby('genre').count()

# count the value of column 'genre' and sort result in descending order 
spr_genres = spr_general.groupby('genre')['track'].count().sort_values(ascending=False)

# show result
spr_genres

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hip            2041
               ... 
metalcore         1
marschmusik       1
malaysian         1
lovers            1
ïîï               1
Name: track, Length: 253, dtype: int64

In [37]:
# show only first 10 rows
spr_genres[:10]

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hip            2041
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: track, dtype: int64

Table above shows that the most popular genre for Springfield listeners is Pop. 

Similar steps will be performed to check if Rap is the most popular genre for Shelbyville listeners. 

First to group table `shel_general` by genre, and find the played tracks count for each genre with `count()`. Results then will be stored in `shel_genres` and displayed in descending order, with the top row being the most popular genre. 

In [38]:
# grouping table 'shel_general' by column genre
shel_general.groupby('genre').count()

# count the value in column genre and display in descending order 
shel_genres = shel_general.groupby('genre')['track'].count().sort_values(ascending=False)
shel_genres

genre
pop           2431
dance         1932
rock          1879
electronic    1736
hip            934
              ... 
mandopop         1
leftfield        1
laiko            1
jungle           1
worldbeat        1
Name: track, Length: 203, dtype: int64

In [39]:
# show only first 10 rows
shel_genres[:10]

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hip             934
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: track, dtype: int64

Table above shows that the most popular genre in Shelbyville is Pop, instead of Rap. 

**Conclusion**

Hypotheses is proven partially correct:
* Pop is the most popular genre in Springfield.  

* However, Rap is NOT the most popular genre in Shelbyville. Rap genre is not even found in top 5 genre list. Listeners in Shelbyville like Pop genre most, which turns out to be similar to listeners in Springfield. 


[Back to Contents](#back)

# General Findings <a id='end'></a>

Three hypoteses in the beginning of this report are as follow:

1. Customer activity differs according to days of the week and city or residence.
2. On Monday AM, listeners from Springfield and Shelbyville listen to different genre. Same applied to Friday PM. 
3. Listeners from Springfield and Shelbyville have different genre preference. 

The data analysis has resulted in conclusions as follow:

1. There is a difference in music listening activities in both cities, depending on days of the week. The first hypothesis is correct. 


2. Music listening activity of Springfield and Shelbyville residents varies little across the week, with only slight difference in genre top 5 ranks between the two cities. They also share the same preference for most popular genre (Pop). Which makes the hypothesis is only partially correct. However this finding is with note that there is a considerably large proportion of missing values ('unknown') in genre for Springfield data. Should this issue be sorted, result may change. 

3. Listeners in both cities share similar music preference, which make the third hypothesis rejected. 

[Back to Contents](#back)